bin/create_testbench_template.py

```diff
@@ -0,0 +1,181 @@
+                input_ports_.append(thing_)
+            elif thing_._port_direction_string == 'GPI_OUTPUT':
+                output_ports_.append(thing_)
+            elif thing_._port_direction_string == 'GPI_INOUT':
+                inout_ports_.append(thing_)
+            else:
+                dut._log.info('Unhandled port direction %s of %s!' % (thing_._port_direction_string, thing_._path))
+
+    # match names to identify "special" ports:
+    for port_ in set(input_ports_+inout_ports_):
+        if port_._name.startswith('vd') or port_._name.startswith('vc'):
+            # Examples: vdd, vddd, vdda, vdio, vcc, vcore
+            dut._log.info('Identified a supply port: %s' % (port_._name))
+            supply_ports_.append(port_)
+        elif port_._name.startswith('vs') or port_._name.startswith('gnd') or port_._name.endswith('gnd'):
+            # Examples: gnd, agnd, dgnd, vs, vss, vssd, vsio
+            dut._log.info('Identified a ground port: %s' % (port_._name))
+            ground_ports_.append(port_)
+        elif (port_._name.startswith('clk') or port_._name.endswith('clk')
+              or port_._name.startswith('clock') or port_._name.endswith('clock')):
+            # Examples: clk_2MHz, clk_main, main_clk, main_clock
+            dut._log.info('Identified a clock: %s' % (port_._name))
+            clock_ports_.append(port_)
+        elif port_._name in ('reset', 'resetn', 'reset_n', 'rst', 'rstn', 'rst_n'):
+            # Examples: exactly these
+            dut._log.info('Identified a reset: %s' % (port_._name))
+            reset_ports_.append(port_)
+
+    # clean up original lists:
+    for port_ in supply_ports_+ground_ports_+clock_ports_+reset_ports_:
+        try:
+            input_ports_.remove(port_)
+        except:
+            pass
+        try:
+            inout_ports_.remove(port_)
+        except:
+            pass
+
+    filename_ = os.getenv('COCOTB_TESTBENCH_TEMPLATE')
+    if filename_ is None:
+        filename_ = 'testbench_template'
+
+    with open(filename_+'.py', 'w') as f:
+        f.write("""# This is a generated testbench template for cocotb (https://cocotb.readthedocs.io)
+
+import cocotb
+from cocotb.triggers import (Timer, Join, RisingEdge, FallingEdge, Edge,
+                             ReadOnly, ReadWrite, ClockCycles, NextTimeStep)
+from cocotb.binary import BinaryValue
+from cocotb.clock import Clock
+from cocotb.result import ReturnValue, TestFailure, TestError, TestSuccess
+
+# TODO: adapt these constants
+CLOCK_PERIOD_NS = 2
+PRE_RESET_TIME_NS = 1
+RESET_DURATION_NS = 7
+
+@cocotb.test()
+def my_first_test(dut):
+    \"\"\"TODO: Add test documentation here.\"\"\"
+
+    import cocotb.wavedrom
+
+    dut._log.info("Running my first test")\n""")
+
+        f.write('\n')
+        f.write('    dut._log.info("Setting ground ports if any were identified")\n')
+        for port_ in ground_ports_:
+            f.write('    dut.%s <= 0\n' % (port_._name))
```

Comment:
This type of formatting looks like a misunderstanding of the guidance to always use tuples when using `%`.

`"%s" % x` is dangerous, because if `x` is a tuple it throws a `TypeError`. The argument goes that you should always use `"%s" % (x,)` so that this problem never comes up.

However, the `"%s" % (x)` form you use here is the worst of both worlds, giving the illusion of addressing this without actually doing so.

You could use `"%s" % (x,)`, but I'd generally recommend the `"{}".format(x)` form, which makes this problem go away (and is closed to the f-strings you originally used)

Reply:
Thanks, I have used ``{}`` now, though with the numbered form for maximum backward compatibility.

Reply:
What are you trying to be compatible with? I sincerely hope cocotb is not trying to support python 2.6 which is 6 years past EOL, and I suspect we are already relying on 2.7 as a minimum anyway.

Reply:
I had the recollection that also some older 3.x version needed this, at least I had to use this format for something unrelated not too long ago.
Cocotb has both styles in its codebase, but I agree we can move on to just ``{}`` here.

Reply:
The documentation claims that 2.6 is still supported... but in practice, it no longer is, since there's a lot of 2.7-exclusive stuff that's been added over the past several months. (I've filed #817 to update the docs).